src/p11_wrap.cpp

```diff
@@ -0,0 +1,274 @@
+    // don't trigger the loading ourselves.
+    auto error = openssl::lib::OpenSSLLib::SSL_ERR_get_error();
+    auto formatter = boost::format("%s: %d");
+    formatter % openssl::lib::OpenSSLLib::SSL_ERR_error_string(error, nullptr) % error;
+    return formatter.str();
+}
+
+/**
+ * This struct is used by CWrap to
+ * determine how to react to an error
+ * when performing a LibP11 call.
+ */
+struct P11ExceptionErrorPolicy
+{
+    template <class Rv>
+    static void handleError(const Rv &);
+};
+
+/**
+ * Throw a P11 exception upon error.
+ *
+ * This method gets the error string and throws an exception
+ * with the corresponding message.
+ */
+template <class Rv>
+void P11ExceptionErrorPolicy::handleError(const Rv & /* unused */)
+{
+    throw P11Exception();
+}
+
+using P11CallPtr =
+        ::cppc::CallCheckContext<::cppc::IsNotNullptrReturnCheckPolicy, P11ExceptionErrorPolicy>;
+using P11IsNonNegative =
+        ::cppc::CallCheckContext<::cppc::IsNotNegativeReturnCheckPolicy, P11ExceptionErrorPolicy>;
+
+P11_SlotInfo::P11_SlotInfo() : _slots(nullptr), _numSlots(0) {}
+
+P11_SlotInfo::P11_SlotInfo(PKCS11_SLOT *slots, unsigned int numSlots)
+        : _slots(slots), _numSlots(numSlots)
+{
+}
+
+static std::string parseID(std::string idHexString)
+{
+    /* We require IDs to be represented as hex strings from the user.
+     * Libp11 does the necessary pre-processing, such as unhexing, on IDs
+     * only when invoked via OpenSSL's Engine API. However, when interacting directly
+     * with LibP11 API, like we do here, this pre-processing is not performed. To make
+     * matters worse, LibP11 does not export its internal
+     * pre-processing function. We therefore begin to re-implement it here.
+     */
+    std::string parsedID = boost::algorithm::unhex(idHexString);
+    return parsedID;
+}
+
+P11_PKCS11_CTX_Ptr _PKCS11_CTX_new(void)
+{
+    return P11_PKCS11_CTX_Ptr{P11CallPtr::callChecked(lib::LibP11::P11_PKCS11_CTX_new)};
+}
+
+void _PKCS11_CTX_load(PKCS11_CTX *ctx, std::string &module)
+{
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_CTX_load, ctx, module.c_str());
+}
+
+void _PKCS11_CTX_unload(PKCS11_CTX *ctx)
+{
+    if (ctx == nullptr) {
+        throw P11Exception("NULL passed for context that is required to be unloaded");
+    }
+    lib::LibP11::P11_PKCS11_CTX_unload(ctx);
+}
+
+P11_SlotInfo _PKCS11_enumerate_slots(PKCS11_CTX *ctx)
+{
+    if (ctx == nullptr) {
+        throw P11Exception("NULL passed for context when enumerating slots");
+    }
+
+    PKCS11_SLOT *slotsp;
+    unsigned int nslotsp;
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_enumerate_slots, ctx, &slotsp, &nslotsp);
+
+    // Based on the information obtained by slot enumaration, we instantiate
+    // a corresponding P11_SlotInfo object as a return value.
+    P11_SlotInfo slotInfo(slotsp, nslotsp);
+    return slotInfo;
+}
+
+void _PKCS11_release_all_slots(PKCS11_CTX *ctx, P11_SlotInfo &slotInfo)
```

Comment:
Why does this function exist? Shouldn't this be the destructor of P11_SlotInfo?
In general, I'm not that happy about the fact that the API allows contexts to be deleted before the respective slotinfo. This seems unsafe. Maybe we should discuss how to avoid problems here.